### DISCUSSION TITLE
Built-in constant `Ellipsis` and ellipsis literal `...` yields different result

### DISCUSSION THREAD

**Discussion by termoshtt** [microsoft/pyright#5669] [Q&A] [answered]
I created two files:

- `test1.py`

```python:test1.py
def f(a=...):
  print(a)

f(1)
f()
```

- `test2.py`

```python:test2.py
def f(a=Ellipsis):
  print(a)

f(1)
f()
```

I expected `pyright` returns same result for these script since `...` and `Ellipsis` are same instance in Python.

https://docs.python.org/3/library/constants.html#Ellipsis
>  Ellipsis
>    The same as the ellipsis literal “...”. 

But `pyright` returns an error only on `test2.py`

```shell
$ pyright test1.py
0 errors, 0 warnings, 0 informations 
$ pyright test2.py
/path/to/test2.py:5:3 - error: Argument of type "Literal[1]" cannot be assigned to parameter "a" of type "ellipsis" in function "f"
    "Literal[1]" is incompatible with "ellipsis" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

Does this intended behavior?

Versions
---------

```shell
$ pyright --version
pyright 1.1.320
$ python --version
Python 3.10.12
$ uname -a
Darwin MyMachineName 22.5.0 Darwin Kernel Version 22.5.0: Thu Jun  8 22:22:23 PDT 2023; root:xnu-8796.121.3~7/RELEASE_ARM64_T6020 arm64
```

**Comment by erictraut** [comment, marked as answer]
Why do you want to use `Ellipsis` rather than the standard `...` syntax? The `...` notation in a function signature has special meaning to a type checker (it evaluates to `Any`), whereas `Ellipsis` does not. I think it's reasonable for a type checker to assume that if a library uses the `Ellipsis` constant in this context, then its intent is different from the standard interpretation.

Python users will also be familiar with `...` and not `Ellipsis`, so it would be best for you to change your code accordingly.

**Reply by termoshtt** [reply]
> Why do you want to use `Ellipsis` rather than the standard `...` syntax?

I encounter this behavior while generating stub file of native extension module generated by [PyO3](https://github.com/pyo3/pyo3) using [inspect.signature](https://docs.python.org/3/library/inspect.html#inspect.signature). Similar script is discussed on https://github.com/PyO3/pyo3/issues/2454#issuecomment-1273300176 . AFAIK, there is no way to tell the argument is ellipsis literal rather than constant in native module since ellipsis literal only appears in AST and Python runtime translates into `Ellipsis` constant immediate after evaluating it.

> The `...` notation in a function signature has special meaning to a type checker (it evaluates to Any), whereas `Ellipsis` does not. I think it's reasonable for a type checker to assume that if a library uses the `Ellipsis` constant in this context, then its intent is different from the standard interpretation.

This is perfect answer for me. I will fix stub file generator to use ellipsis literal. Thanks 😄
